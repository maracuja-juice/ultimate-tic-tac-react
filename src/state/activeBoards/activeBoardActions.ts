import { Point } from '../../util/Point';

export const SET_ACTIVE_BOARDS = 'activeBoardReducer/SetActiveBoard';
export const CALCULATE_ACTIVE_BOARDS = 'activeBoardReducer/CalculateActiveBoard';

export const setActiveBoards = ( boards: Point[], ) => ({
    type: SET_ACTIVE_BOARDS,
    payload: boards
});
export const calculateActiveBoards = ( lastMove: Point ) => ({
    type: CALCULATE_ACTIVE_BOARDS,
    payload: lastMove
});