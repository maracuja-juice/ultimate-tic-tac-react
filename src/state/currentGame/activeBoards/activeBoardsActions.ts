import { Point } from '../../../util/Point';

export const SET_ALLOWED_BOARDS = 'activeBoardsReducer/set-allowed-boards';
export const CALCULATE_ALLOWED_BOARDS = 'activeBoardsReducer/calculate-allowed-boards';

export const setAllowedBoards = ( boards: Point[], ) => ({
    type: SET_ALLOWED_BOARDS,
    payload: boards
});
export const calculateActiveBoards = ( lastMove: Point ) => ({
    type: CALCULATE_ALLOWED_BOARDS,
    payload: lastMove
});