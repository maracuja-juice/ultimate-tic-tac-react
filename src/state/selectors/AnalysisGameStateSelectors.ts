import { AnalysisGame, AppState, Player } from '../AppState';

export const getAnalysisGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      id: finishedGame.id!,
      board: finishedGame.gameState,
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
    };
  }

  return undefined;
};
