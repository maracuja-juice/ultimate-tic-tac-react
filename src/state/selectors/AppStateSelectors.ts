import { AnalysisGame, AppState, Player } from '../AppState';

export const getBoards = (state: AppState) => state.currentGame.board;

export const getCurrentPlayer = (state: AppState) =>
  state.currentGame.game.currentPlayer;

export const getWinningPlayer = (state: AppState) =>
  state.currentGame.game.winningPlayer;

export const getMoves = (state: AppState) => state.currentGame.moves;

export const getFinishedGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
      board: finishedGame.gameState,
      id: finishedGame.id!,
    };
  }

  return undefined;
};
