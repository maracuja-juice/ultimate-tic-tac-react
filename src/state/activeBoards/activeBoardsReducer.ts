import { Point } from '../../util/Point';
import { GenericAction } from '../AppState';
import { SET_ALLOWED_BOARDS } from './activeBoardsActions';
import { RESTART_GAME } from '../commonAction';

const getAllPoints = () => {
    let points: Point[] = [];

    for (let x = 0; x < 3; x++) {
        for (let y = 0; y < 3; y++) {
            points.push( {x, y} );
        }
    }

    return points;
};

const initialState: Point[] = getAllPoints();

const activeBoardsReducer = ( state = initialState, action: GenericAction ) => {
    switch (action.type) {
        case SET_ALLOWED_BOARDS: {
            return action.payload;
        }
        case RESTART_GAME: {
            return initialState;
        }
        default: {
            return state;
        }
    }
};

export default activeBoardsReducer;