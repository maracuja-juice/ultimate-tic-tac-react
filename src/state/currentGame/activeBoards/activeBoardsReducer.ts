import { Point, TicTacToeGame } from '../../../lib';
import { GenericAction } from '../../AppState';
import { SET_ACTIVE_BOARDS } from './activeBoardsActions';
import { RESTART_GAME } from '../../commonAction';

const initialState: Point[] = new TicTacToeGame([]).getCurrentActiveBoards();

const activeBoardsReducer = (state = initialState, action: GenericAction) => {
  switch (action.type) {
    case SET_ACTIVE_BOARDS: {
      return action.payload;
    }
    case RESTART_GAME: {
      return initialState;
    }
    default: {
      return state;
    }
  }
};

export default activeBoardsReducer;
