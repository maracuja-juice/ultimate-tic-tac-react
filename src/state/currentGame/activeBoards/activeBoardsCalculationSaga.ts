import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction } from '../../AppState';
import { CALCULATE_ALLOWED_BOARDS, setAllowedBoards } from './activeBoardsActions';
import { getBoards } from '../../selectors/AppStateSelectors';
import { getNewActiveBoards } from '../../../util/ActiveBoards';

function* calculateActiveBoards( action: GenericAction ) {
    const boards = yield select( getBoards );
    const lastMove = action.payload;

    const activeBoards = getNewActiveBoards(lastMove, boards);

    yield put( setAllowedBoards( activeBoards ) );
}

function* activeBoardsCalculationSaga() {
    yield takeEvery( CALCULATE_ALLOWED_BOARDS, calculateActiveBoards );
}

export default activeBoardsCalculationSaga;