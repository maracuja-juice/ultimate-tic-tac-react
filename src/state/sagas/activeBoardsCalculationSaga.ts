import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction, SmallBoardInformation, TileValue } from '../AppState';
import { arePointsEqual } from '../../util';
import { CALCULATE_ALLOWED_BOARDS, setAllowedBoards } from '../activeBoards/activeBoardActions';
import { getBoards } from '../selectors/AppStateSelectors';

function* calculateActiveBoards( action: GenericAction ) {
    const boards = yield select( getBoards );
    const lastMove = action.payload;

    let activeBoards = [lastMove];
    const boardLastMovePointsTo = boards.find( ( board: SmallBoardInformation ) =>
                                                   arePointsEqual( board.position, lastMove ) );
    const boardIsFinished = boardLastMovePointsTo.value !== TileValue.Empty;
    if (boardIsFinished) {
        const allUnfinishedBoards = boards.filter(
            ( board: SmallBoardInformation ) =>
                board.value === TileValue.Empty
        );
        activeBoards = allUnfinishedBoards.map( ( board: SmallBoardInformation ) => board.position );
    }

    yield put( setAllowedBoards( activeBoards ) );
}

function* activeBoardsCalculationSaga() {
    yield takeEvery( CALCULATE_ALLOWED_BOARDS, calculateActiveBoards );
}

export default activeBoardsCalculationSaga;