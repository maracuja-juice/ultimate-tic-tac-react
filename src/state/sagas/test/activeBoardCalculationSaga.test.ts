import { expectSaga } from 'redux-saga-test-plan';
import { select } from 'redux-saga/effects';
import { getBoards } from '../../selectors/AppStateSelectors';
import { CALCULATE_ALLOWED_BOARDS, SET_ALLOWED_BOARDS } from '../../activeBoards/activeBoardActions';
import activeBoardsCalculationSaga from '../activeBoardsCalculationSaga';
import unfinishedBoardMock from './unfinishedBoardMock';
import { circleFinishedBoardMock } from './finishedBoardMock';

describe( 'activeBoardCalculationSaga', () => {
    it( 'should dispatch setAllowedBoard Action, normal case', () => {
        return expectSaga( activeBoardsCalculationSaga )
            .provide( [
                          [select( getBoards ), circleFinishedBoardMock]
                      ] )
            .put( {type: SET_ALLOWED_BOARDS, payload: [{x: 2, y: 0}]} )
            .dispatch( {type: CALCULATE_ALLOWED_BOARDS, payload: {x: 2, y: 0}} )
            .run();
    } );

    it( 'should set allowedBoards to all unfinished small boards ' +
        'when small board the last move points to is finished',
        () => {
            // these values depend on the unfinished board mock.
            return expectSaga( activeBoardsCalculationSaga )
                .provide( [
                              [select( getBoards ), unfinishedBoardMock]
                          ] )
                .put( {
                          type: SET_ALLOWED_BOARDS,
                          payload: [{x: 0, y: 2}, {x: 1, y: 0}, {x: 1, y: 2}, {x: 2, y: 0}, {x: 2, y: 2}]} )
                .dispatch( {type: CALCULATE_ALLOWED_BOARDS, payload: {x: 2, y: 1}} )
                .run();
        } );

    // if more put effects happen: this catches it + this checks for the order
    it( 'should match snapshot', () => {
        return expectSaga( activeBoardsCalculationSaga )
            .provide( [
                          [select( getBoards ), circleFinishedBoardMock]
                      ] )
            .put( {type: SET_ALLOWED_BOARDS, payload: [{x: 2, y: 0}]} )
            .dispatch( {type: CALCULATE_ALLOWED_BOARDS, payload: {x: 2, y: 0}} )
            // TODO maybe this needs to be switched to 0,2
            .run()
            .then( ( result ) => {
                expect( result.toJSON() ).toMatchSnapshot();
            } );
    } );
} );
