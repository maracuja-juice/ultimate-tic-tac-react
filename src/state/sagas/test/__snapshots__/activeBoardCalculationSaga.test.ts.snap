// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activeBoardCalculationSaga should match snapshot 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "activeBoardReducer/calculate-allowed-boards",
          "@@redux-saga-test-plan/json/function/calculateActiveBoards",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          Object {
            "payload": Object {
              "x": 2,
              "y": 0,
            },
            "type": "activeBoardReducer/calculate-allowed-boards",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "activeBoardReducer/calculate-allowed-boards",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "activeBoardReducer/calculate-allowed-boards",
      },
    },
  ],
}
`;
